File: app/views/assets.py
# -*- coding: utf-8 -*-
from app.views.common import *

# 定义蓝图
assets = Blueprint('assets', __name__)


# 固定资产主机列表HTML模块
@assets.route('/hosts', methods=["GET", "POST"])
@login_required  # 登录保护
def hosts():
    return render_template("assets/hosts.html")


# 固定资产主机模块
@assets.route('/hostsmg', methods=["GET", "POST"])
@login_required  # 登录保护
@user_jurisdiction
def hostsmg():
    if request.method == 'POST':
        action = request.form['action']
        if action == "list":
            try:
                hostlist = []
                for host in db.session.query(Hosts).all():
                    hostlist.append({"id": host.id, "hostname": host.hostname, "os": host.os, "osrelease": host.osrelease,
                                     "kernelrelease": host.kernelrelease, "selinux": host.selinux
                                        , "status": host.status, "mem_total": host.mem_total, "num_cpus": host.num_cpus,
                                     "add_time": host.add_time, "up_time": host.up_time, "eth0_ipaddr": host.eth0_ipaddr
                                     })
                return json.dumps({"code": 1, "msg": u"请求数据成功!", "data": hostlist}, cls=MyEncoder)
            except Exception as error:
                print(str(error))
                return json.dumps({"code": 1, "msg": u"请求数据失败!", "data": ""}, cls=MyEncoder)

        if action == "hostsup":
            try:
                 # 初始化saltstack api
                 salt = saltAPI(host=app.config['SALT_API_ADDR'], user=app.config['SALT_API_USER'],
                                password=app.config['SALT_API_USER'], prot=app.config['SALT_API_PROT'])

                 SaltHostAll = salt.saltCmd({"fun": "key.list_all", "client": "wheel"})[0]

                 # 删除不存在的主机
                 for host in db.session.query(Hosts).all():
                     if not host.hostname in SaltHostAll["data"]["return"]["minions"]:
                         db.session.delete(Hosts.query.get(host.id))
                         db.session.commit()

                 HostAllInfo = salt.saltCmd({"fun": "grains.items", "client": "local", "tgt": "*"})[0]  # 获取所有主机信息
                 for host in HostAllInfo:
                     try:
                         ipaddr = HostAllInfo[host]['ip4_interfaces']['eth0'][0]
                     except Exception as error:
                         print(str(error))
                         ipaddr = ""
                     # 查询当前主机信息
                     HostInfo = db.session.query(Hosts).filter_by(hostname=host).first()
                     if HostInfo is not None and HostAllInfo[host] == False:  # db主机存在 且 salt返回 false；及主机离线
                         HostInfo.status = "0"
                     elif HostInfo is None and HostAllInfo[host] != False:  # DB主机不存在且salt 返回不为 false；及新增主机
                         HostData = mysqld.Hosts(
                             hostname=HostAllInfo[host]['id'],
                             os=HostAllInfo[host]['os'],
                             osrelease=HostAllInfo[host]['osrelease'],
                             kernelrelease=HostAllInfo[host]['kernelrelease'],
                             selinux=HostAllInfo[host]['selinux']['enforced'],
                             status="1",
                             mem_total=HostAllInfo[host]['mem_total'],
                             num_cpus=HostAllInfo[host]['num_cpus'],
                             eth0_ipaddr=ipaddr
                         )
                         db.session.add(HostData)
                     elif HostInfo is not None and HostAllInfo[host] != False:  # DB主机存在且 salt 返回不为false ； 及更新主机信息
                         HostInfo.os = HostAllInfo[host]['os']
                         HostInfo.osrelease = HostAllInfo[host]['osrelease']
                         HostInfo.kernelrelease = HostAllInfo[host]['kernelrelease']
                         HostInfo.selinux = HostAllInfo[host]['selinux']['enforced']
                         HostInfo.status = "1"
                         HostInfo.mem_total = HostAllInfo[host]['mem_total']
                         HostInfo.num_cpus = HostAllInfo[host]['num_cpus']
                         HostInfo.up_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
                         HostInfo.eth0_ipaddr = ipaddr
                     db.session.commit()
                 logs("hostsmg.hostsup",u"更新主机列表成功")
                 return json.dumps({"code": 1, "msg": u"更新主机列表成功!", "data": ""}, cls=MyEncoder)
            except Exception as error:
                print(str(error))
                return json.dumps({"code": -1, "msg": u"更新主机列表失败!", "data": ""}, cls=MyEncoder)
        if action == "hostinfo":
            try:
                HostName = request.form['hostname']
                salt = saltAPI(host=app.config['SALT_API_ADDR'], user=app.config['SALT_API_USER'],
                           password=app.config['SALT_API_USER'], prot=app.config['SALT_API_PROT'])
                HostAllInfo = salt.saltCmd({"fun": "grains.items", "client": "local", "tgt": HostName})[0]
                return json.dumps({"code": 1, "msg": u"请求数据成功!", "data": Json_Unicode_To_Uft8(HostAllInfo)}, cls=MyEncoder)
            except Exception as error:
                print(str(error))
                return json.dumps({"code": -1, "msg": u"请求数据成功!", "data": ""}, cls=MyEncoder)

File: app/static/js/assets.js
//更新固定资产主机列表
function assets_update_hosts() {
    layer.msg("更新中，请稍后....", {time: 0});
    $.ajax({
        type: 'post',
        url: '/assets/hostsmg',
        data: {
            "action": 'hostsup'
        },
        dataType: 'json',
        success: function (data) {
            if (data.code == 1) {
                layer.msg(data.msg);
                assets_hosts_list()
            } else {
                layer.msg(data.msg)
            }
        }
    });
}


//显示主机详细信息
function hostsinfo(hostname) {
    layer.msg("数据加载中，请稍后....", {time: 0});
    $.ajax({
        type: 'post',
        url: '/assets/hostsmg',
        data: {
            "action": 'hostinfo',
            "hostname": hostname
        },
        dataType: 'json',
        success: function (res) {
            layer.msg("数据加载完成!");
            if (res.code == 1) {
                layer.open({
                    type: 1,
                    //skin: 'layui-layer-rim', //加上边框
                    area: ['820px', '840px'], //宽高
                    content: '<pre id="result">' + syntaxHighlight(res.data) + '</pre>'
                });
            } else {
                layer.mgs(res.mgs)
            }
        }
    });
}


function assets_hosts_list() {
    $.ajax({
        type: 'post',
        url: '/assets/hostsmg',
        data: {
            "action": 'list'
        },
        dataType: 'json',
        success: function (js) {
            var obj = js;
            if (obj.code != 1) {
                layer.msg(obj.msg)
            } else {
                var trStr = '';
                for (i = 0; i < obj.data.length; i++) {
                    trStr += '<tr> ';
                    trStr += '<td><p style="font-weight:bold">' + obj.data[i].hostname + '</p></td> ';
                    trStr += '<td> ' + obj.data[i].os + obj.data[i].osrelease + '</td>';
                    trStr += '<td> ' + obj.data[i].kernelrelease + '</td>';
                    trStr += '<td> ' + obj.data[i].eth0_ipaddr + '</td>';
                    trStr += '<td> ' + obj.data[i].selinux + '</td>';
                    trStr += '<td> ' + obj.data[i].num_cpus + '</td>';
                    trStr += '<td> ' + obj.data[i].mem_total + '</td>';
                    if (obj.data[i].status == "1") {
                        trStr += '<td><a><span class="label label-success">正常</span></a></td>';
                    } else if (obj.data[i].status == "0") {
                        trStr += '<td><a><span class="label label-danger">离线</span></a>'
                    } else {
                        trStr += '<td><a><span class="label label-warning">未知</span></a>'
                    }
                    trStr += '<td> ' + obj.data[i].add_time + '</td>';
                    trStr += '<td> ' + obj.data[i].up_time + '</td>';
                    trStr += '<td onclick=hostsinfo("' + obj.data[i].hostname + '")><a href="#"> <span class="label label-info" ><i class="fa fa-fw fa-bars"></i>详细</span></a></td>';
                    trStr += '</tr> ';
                }
                $("#assets-hosts-list").html(trStr);
            }
        }
    });

}